doc/source/data/examples/custom_datasource_example.rst

```diff
@@ -0,0 +1,107 @@
+Optionally, specify valid file extensions. Ray Data ignores files with other extensions.
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __datasource_constructor_start__
+    :end-before: __datasource_constructor_end__
+
+Implement ``_read_stream``
+==========================
+
+``_read_stream`` is a generator that yields one or more blocks of data from a file. 
+
+Blocks are PyArrow tables, pandas DataFrames, or dictionaries of NumPy arrays. Don't 
+create a block directly. Instead, add rows of data to a  ``DelegatingBlockBuilder``.
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __read_stream_start__
+    :end-before: __read_stream_end__
+
+Read your data
+==============
+
+Once you've implemented ``ImageDatasource``, call :func:`~ray.data.read_datasource` to 
+read images into a :class:`~ray.data.Dataset`. Ray Data reads your files in parallel.
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __read_datasource_start__
+    :end-before: __read_datasource_end__
+
+Write data to files
+-------------------
+
+The core abstractions for writing data to files are :class:`~ray.data.datasource.RowBasedFileDatasink` and 
+:class:`~ray.data.datasource.BlockBasedFileDatasink`. They provide file-specific functionality on top of the
+:class:`~ray.data.Datasink` interface.
+
+If you want to write one row per file, subclass :class:`~ray.data.datasource.RowBasedFileDatasink`. 
+Otherwise, subclass :class:`~ray.data.datasource.BlockBasedFileDatasink`.
+
+.. vale off
+
+.. Ignoring Vale because of future tense.
+
+In this example, you'll write one image per file, so you'll subclass 
+:class:`~ray.data.datasource.RowBasedFileDatasink`. To subclass 
+:class:`~ray.data.datasource.RowBasedFileDatasink`, implement the constructor and 
+:meth:`~ray.data.datasource.RowBasedFileDatasink.write_row_to_file`.
+
+.. vale on
+
+Implement the constructor
+=========================
+
+Call the superclass constructor and specify the folder to write to. Optionally, specify
+a string representing the file format (for example, ``"png"``). Ray Data uses the
+file format as the file extension.
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __datasink_constructor_start__
+    :end-before: __datasink_constructor_end__
+
+Implement ``write_row_to_file``
+===============================
+
+``write_row_to_file`` writes a row of data to a file. Each row is a dictionary that maps
+column names to values. 
+
+.. literalinclude:: ../doc_code/custom_datasource_example.py
+    :language: python
+    :start-after: __write_row_to_file_start__
+    :end-before: __write_row_to_file_end__
+
+Write your data
+===============
+
+Once you've implemented ``ImageDatasink``, call :meth:`~ray.data.Dataset.write_datasink`
```

Comment:
Same thing here, show how to pass args through.

Reply:
This example already passes two user-specified arguments (`column` and `file_format`)